### PR TITLE
Feature/resumption during low voltage

### LINKED ISSUE #2233: [SDL 0170] SDL behavior in case of LOW_VOLTAGE event
Proposal: [SDL behavior in case of LOW_VOLTAGE event](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0170-sdl-behavior-in-case-of-Low-Voltage.md)

> This proposal defines SDL behavior when HMI requires SDL's services to be suspended due to a low voltage event. A low voltage scenario is triggered when the battery voltage hits below a certain predefined threshold set by the system. In case of such an event, the EMMC is turned off and all read/write operations are unavailable. After the voltage level is restored all operations are resumed.

Review: smartdevicelink/sdl_evolution#489

Steering Committee Decision:
> The Steering Committee has voted to accept this proposal with revisions. The revisions will include the use of 3 signals of the range SIGRTMIN - SIGRTMAX for LowVoltage, WakeUp and IgnitionOff - in place of using mqueue messages.

> The proposal .md was updated on June 5, 2018 to include the agreed upon revisions.

### PR DESCRIPTION
**Pull request by AByzhynar**
Fixes #2233

This PR is **[ready]** for review.

### Risk
This PR makes **[no]** API changes.

### Testing Plan
Unit tests will be provided
ATF tests will be provided

### Summary
Added resumption logic during Low Voltage event

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: src/components/application_manager/include/application_manager/resumption/resume_ctrl_impl.h [hunk 1]
```diff
@@ -138,6 +139,30 @@ class ResumeCtrlImpl : public ResumeCtrl,
    */
   void OnAwake() OVERRIDE;
 
+  /**
+   * @brief Returns Low Voltage signal timestamp
+   * @return Low Voltage event timestamp if event LOV VOLTAGE event occures
+   * otherwise 0
+   */
+  time_t LowVoltageTime() const OVERRIDE;
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/include/application_manager/resumption/resume_ctrl_impl.h]: @AByzhynar please avoid duplication of comments in implementation and interface

File: src/components/application_manager/include/application_manager/resumption/resume_ctrl.h [hunk 2]
```diff
@@ -124,6 +124,40 @@ class ResumeCtrl {
...
+
+  /**
+   * @brief Saves Wake Up signal timestamp
+   */
+  virtual void SaveWakeUpTime() = 0;
+
+  /**
+   * @brief Resets Low Voltage signal timestamp to zero
+   */
+  virtual void ResetLowVoltageTime() = 0;
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/include/application_manager/resumption/resume_ctrl.h]: @AByzhynar check if this functions should be public in interface

File: src/components/application_manager/include/application_manager/resumption/resume_ctrl_impl.h [hunk 3]
```diff
@@ -390,11 +432,30 @@ class ResumeCtrlImpl : public ResumeCtrl,
   void AddWayPointsSubscription(app_mngr::ApplicationSharedPtr application,
                                 const smart_objects::SmartObject& saved_app);
 
+  /**
+   * @brief Checks if saved HMI level is allowed for resumption
+   * by Ignition Cycle restrictions
+   * @param saved_app application specific section from backup file
+   * @return True is allowed , othrwise - False
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/include/application_manager/resumption/resume_ctrl_impl.h]: @AByzhynar typo
`true if allowed`

File: src/components/application_manager/src/resumption/resume_ctrl_impl.cc [hunk 4]
```diff
@@ -717,21 +765,62 @@ void ResumeCtrlImpl::AddSubscriptions(
...
     LOG4CXX_INFO(logger_, "Application was connected long after ign on");
-    result = false;
+    return false;
   }
 
-  if (!DisconnectedJustBeforeIgnOff(saved_app)) {
-    LOG4CXX_INFO(logger_, "Application was dissconnected long before ign off");
-    result = false;
+  if (!CheckDelayBeforeIgnOff(saved_app)) {
+    LOG4CXX_INFO(logger_, "Application was disconnected long before ign off");
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/resumption/resume_ctrl_impl.cc]: @AByzhynar rename to debug

File: src/components/application_manager/include/application_manager/resumption/resume_ctrl_impl.h (hunk 1 above)
Comment (AByzhynar) [reply src/components/application_manager/include/application_manager/resumption/resume_ctrl_impl.h]: @LuxoftAKutsan Done. Sorry, commit squashed to old one.

File: src/components/application_manager/include/application_manager/resumption/resume_ctrl.h (hunk 2 above)
Comment (AByzhynar) [reply src/components/application_manager/include/application_manager/resumption/resume_ctrl.h]: @LuxoftAKutsan Done. Sorry, commit squashed to old one.

File: src/components/application_manager/include/application_manager/resumption/resume_ctrl_impl.h (hunk 3 above)
Comment (AByzhynar) [reply src/components/application_manager/include/application_manager/resumption/resume_ctrl_impl.h]: @LuxoftAKutsan Done. Sorry, commit squashed to old one.

File: src/components/application_manager/src/application_manager_impl.cc [hunk 5]
```diff
@@ -2612,10 +2613,12 @@ void ApplicationManagerImpl::UnregisterApplication(
 
       return;
     }
-    if (is_resuming) {
-      resume_controller().SaveApplication(app_to_remove);
-    } else {
-      resume_controller().RemoveApplicationFromSaved(app_to_remove);
+    if (!IsLowVoltage()) {
```
[the remark this answers is not in the document]
Comment (AByzhynar) [reply src/components/application_manager/src/application_manager_impl.cc]: @LuxoftAKutsan Done. Sorry, commit squashed to old one.

File: src/components/application_manager/src/resumption/resume_ctrl_impl.cc (hunk 4 above)
Comment (AByzhynar) [reply src/components/application_manager/src/resumption/resume_ctrl_impl.cc]: @LuxoftAKutsan Done. Sorry, commit squashed to old one.

File: src/components/application_manager/src/resumption/resumption_data_json.cc [hunk 6]
```diff
@@ -283,6 +284,60 @@ uint32_t ResumptionDataJson::GetIgnOffTime() const {
   return resumption[strings::last_ign_off_time].asUInt();
 }
 
+uint32_t ResumptionDataJson::GetGlobalIgnOnCounter() const {
+  using namespace app_mngr;
+  LOG4CXX_AUTO_TRACE(logger_);
+  sync_primitives::AutoLock autolock(resumption_lock_);
+  Json::Value& resumption = GetResumptionData();
+  if (!resumption.isMember(strings::global_ign_on_counter)) {
+    resumption[strings::global_ign_on_counter] = 1;
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/resumption/resumption_data_json.cc]: @AByzhynar it is not expected that getter will change data.

File: src/components/application_manager/src/resumption/resumption_data_json.cc [hunk 7]
```diff
@@ -283,6 +284,60 @@ uint32_t ResumptionDataJson::GetIgnOffTime() const {
...
+  return resumption[strings::global_ign_on_counter].asUInt();
+}
+
+uint32_t ResumptionDataJson::GetGlobalIgnOffCounter() const {
+  using namespace app_mngr;
+  LOG4CXX_AUTO_TRACE(logger_);
+  sync_primitives::AutoLock autolock(resumption_lock_);
+  Json::Value& resumption = GetResumptionData();
+  if (!resumption.isMember(strings::global_ign_off_counter)) {
+    resumption[strings::global_ign_off_counter] = 0;
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/resumption/resumption_data_json.cc]: @AByzhynar it is not expected that getter will change data.

File: src/components/application_manager/src/resumption/resumption_data_json.cc (hunk 6 above)
Comment (AByzhynar) [reply src/components/application_manager/src/resumption/resumption_data_json.cc]: @LuxoftAKutsan Absolutely agree

File: src/components/application_manager/include/application_manager/resumption/resume_ctrl_impl.h [hunk 8]
```diff
@@ -309,6 +321,30 @@ class ResumeCtrlImpl : public ResumeCtrl,
...
+   * otherwise 0
+   */
+  time_t LowVoltageTime() const;
+
+  /**
+   * @brief Returns Wake Up signal timestamp
+   * @return Wake Up timestamp if Wake Up signal occures
+   * otherwise 0
+   */
+  time_t WakeUpTime() const;
```
[the remark this answers is not in the document]
Comment (AByzhynar) [reply src/components/application_manager/include/application_manager/resumption/resume_ctrl_impl.h]: @AKalinich-Luxoft Disagree. it is the same type for all events. I don't see any reason to complicate the code.

File: src/components/application_manager/src/resumption/resume_ctrl_impl.cc [hunk 9]
```diff
@@ -753,6 +829,104 @@ bool ResumeCtrlImpl::DisconnectedJustBeforeIgnOff(
...
+  LOG4CXX_AUTO_TRACE(logger_);
+  DCHECK_OR_RETURN(saved_app.keyExists(strings::time_stamp), false);
+
+  const time_t time_stamp =
+      static_cast<time_t>(saved_app[strings::time_stamp].asInt());
+  time_t ign_off_time =
+      static_cast<time_t>(resumption_storage_->GetIgnOffTime());
+
+  if (0 == ign_off_time) {
+    return true;
```
[the remark this answers is not in the document]
Comment (AByzhynar) [reply src/components/application_manager/src/resumption/resume_ctrl_impl.cc]: @LuxoftAKutsan Done in 93ff02d145e8fd2ab3f2ab263d75b6d229fe34d6

File: src/components/application_manager/src/resumption/resumption_data_json.cc (hunk 6 above)
Comment (AByzhynar) [reply src/components/application_manager/src/resumption/resumption_data_json.cc]: @LuxoftAKutsan Done in 93ff02d145e8fd2ab3f2ab263d75b6d229fe34d6

File: src/components/application_manager/src/resumption/resumption_data_json.cc (hunk 7 above)
Comment (AByzhynar) [reply src/components/application_manager/src/resumption/resumption_data_json.cc]: @LuxoftAKutsan Done in 93ff02d145e8fd2ab3f2ab263d75b6d229fe34d6

File: src/components/application_manager/include/application_manager/resumption/resume_ctrl.h [hunk 10]
```diff
@@ -124,6 +124,40 @@ class ResumeCtrl {
    */
   virtual void OnAwake() = 0;
 
+  /**
+   * @brief Returns Low Voltage signal timestamp
+   * @return Low Voltage event timestamp if event LOV VOLTAGE event occures
```
[the remark this answers is not in the document]
Comment (AByzhynar) [reply src/components/application_manager/include/application_manager/resumption/resume_ctrl.h]: @LuxoftAKutsan Done in 93ff02d145e8fd2ab3f2ab263d75b6d229fe34d6

File: src/components/application_manager/include/application_manager/resumption/resume_ctrl_impl.h [hunk 11]
```diff
@@ -409,11 +464,36 @@ class ResumeCtrlImpl : public ResumeCtrl,
...
   bool CheckDelayAfterIgnOn();
 
+  /**
+   * @brief CheckDelayBeforeIgnOff checks if app was unregistered less
+   * then N seconds before Ignition OFF. N will be read from profile.
+   * @return true if app was disconnected within timeframe of N seconds before
+   * Ignition Off,
+   * otherwise return false
+   */
+  bool CheckDelayBeforeIgnOff(const smart_objects::SmartObject& saved_app);
```
[the remark this answers is not in the document]
Comment (AByzhynar) [reply src/components/application_manager/include/application_manager/resumption/resume_ctrl_impl.h]: @LuxoftAKutsan Done in 93ff02d145e8fd2ab3f2ab263d75b6d229fe34d6

File: src/components/application_manager/src/resumption/resume_ctrl_impl.cc [hunk 12]
```diff
@@ -789,11 +964,32 @@ bool ResumeCtrlImpl::CheckIcons(ApplicationSharedPtr application,
...
 }
 
+bool ResumeCtrlImpl::CheckIgnCyclesData() const {
+  LOG4CXX_AUTO_TRACE(logger_);
+  const uint32_t global_ign_off_count =
+      resumption_storage_->GetGlobalIgnOffCounter();
+  const uint32_t global_ign_on_count =
+      resumption_storage_->GetGlobalIgnOnCounter();
+  const uint32_t diff = global_ign_on_count - global_ign_off_count;
+  if (diff >= 2) {
```
[the remark this answers is not in the document]
Comment (AByzhynar) [reply src/components/application_manager/src/resumption/resume_ctrl_impl.cc]: @LuxoftAKutsan Done in 93ff02d145e8fd2ab3f2ab263d75b6d229fe34d6

File: src/components/application_manager/src/resumption/resume_ctrl_impl.cc [hunk 13]
```diff
@@ -335,6 +347,37 @@ void ResumeCtrlImpl::OnAwake() {
...
+void ResumeCtrlImpl::SaveLowVoltageTime() {
+  ResetLowVoltageTime();
+  low_voltage_time_ = time(nullptr);
+  LOG4CXX_DEBUG(logger_,
+                "Low Voltage timestamp : " << low_voltage_time_ << " saved");
+}
+
+void ResumeCtrlImpl::ResetLowVoltageTime() {
+  low_voltage_time_ = 0;
+  LOG4CXX_DEBUG(logger_, "Resetting Low Voltage timestamp :");
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/resumption/resume_ctrl_impl.cc]: @AByzhynar redundant semicolon

File: src/components/application_manager/src/resumption/resume_ctrl_impl.cc [hunk 14]
```diff
@@ -335,6 +347,37 @@ void ResumeCtrlImpl::OnAwake() {
...
+
+void ResumeCtrlImpl::SaveWakeUpTime() {
+  ResetWakeUpTime();
+  wake_up_time_ = std::time(nullptr);
+  LOG4CXX_DEBUG(logger_, "Wake Up timestamp : " << wake_up_time_ << " saved");
+}
+
+void ResumeCtrlImpl::ResetWakeUpTime() {
+  wake_up_time_ = 0;
+  LOG4CXX_DEBUG(logger_, "Resetting Wake Up timestamp :");
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/resumption/resume_ctrl_impl.cc]: @AByzhynar redundant semicolon

File: src/components/application_manager/src/resumption/resume_ctrl_impl.cc (hunk 13 above)
Comment (AByzhynar) [reply src/components/application_manager/src/resumption/resume_ctrl_impl.cc]: @LuxoftAKutsan Done and squashed.

File: src/components/application_manager/src/resumption/resume_ctrl_impl.cc (hunk 14 above)
Comment (AByzhynar) [reply src/components/application_manager/src/resumption/resume_ctrl_impl.cc]: @LuxoftAKutsan Done and squashed.

**Pull request merged**